lib/puppet/application/lookup.rb

```diff
@@ -365,9 +353,31 @@ def generate_scope
       unless given_facts.instance_of?(Hash)
         raise _("Incorrect formatted data in %{fact_file} given via the --facts flag") % { fact_file: fact_file }
       end
-      node.add_extra_facts(given_facts)
     end
 
+    extra_facts = Puppet::Node::Facts.indirection.find(options[:node], :environment => Puppet.lookup(:current_environment))
+    if extra_facts && given_facts
+      extra_facts.add_extra_values(given_facts)
+    elsif given_facts
+      extra_facts = Puppet::Node::Facts.new(node.name, given_facts)
+    else
+      raise _("No facts available for target node: %{node}") % { node: node.name}
+    end
```

Comment:
It is possible that the `#find` method from the indirector will return `nil`, if there are no cached facts on the node, so `#add_extra_values` would be called by a `nil` object. This implementation fixes this, and also covers the fact file case, but breaks a lot of tests. (also just a bad implementation overall).
Also not sure about `Puppet.lookup(:current_environment)`.